### DISCUSSION TITLE
Adding extra setup files using the vitest API

### DISCUSSION THREAD

**Discussion by odinvanderlinden** [vitest-dev/vitest#3015] [Q&A] [answered]
To support vitest as a test runner for [stryker](https://github.com/stryker-mutator/stryker-js/issues/3465) we would like to add extra setup files next to the setup files the user has declared. We came up with this solution:
![image](https://user-images.githubusercontent.com/45065297/225399454-32b51418-1751-48a3-9b47-bfb36f604648.png)
We were wondering if this is the correct way to use the API? If so would you guys allow us to add it to the vitest docs?

**Comment by sheremet-va** [comment, marked as answer]
Can't you just pass it down inside a second argument? Or does it override the user's setup files?

**Reply by nicojs** [reply]
It indeed overrides the user setup. We (@odinvanderlinden mainly 😅) are actually implementing the [StrykerJS](stryker-mutator.github.io) plugin for Vitest, where we need to extend the existing user config and definitely not override it.

Our company (Info Support) is also a sponsor of you and @antfu, maybe that helps 😅

**Reply by sheremet-va** [reply]
> where we need to extend the existing user config and definitely not override it

Then your solution is probably a good one and future-proof. The problem with modifying the config is that its resolution is quite complex and can depend on user plugins (`config.plugins`), so we need to run it through Vite's plugin pipeline.

> Our company (Info Support) is also a sponsor of you and @antfu, maybe that helps 😅

Oh, thank you very much for your support 😄

**Reply by nicojs** [reply]
> Then your solution is probably a good one and future-proof.

So you would be ok with a PR to add this use case to this page? https://vitest.dev/advanced/api.html 

> The problem with modifying the config is that its resolution is quite complex and can depend on user plugins

Yes, that's what we figured as well. We were very happy to see our approach  worked.

You're welcome. Thank you for maintaining vitest!

**Reply by sheremet-va** [reply]
> So you would be ok with a PR to add this use case to this page? [vitest.dev/advanced/api.html](https://vitest.dev/advanced/api.html)

Yes, contribution is welcome 😄

**Reply by sheremet-va** [reply]
But I would still like to improve the API. I have a feeling that your solution might not work with an experimental optimizer. 👀

**Reply by nicojs** [reply]
What is an experimental optimizer and why wouldn't adding setup files work for that approach? 😅

**Comment by sheremet-va** [comment]
> We were wondering if this is the correct way to use the API? If so would you guys allow us to add it to the vitest docs?


The public API is not standardised. We can extend it based on the feedback if needed.
